[`Whiper`] add `get_input_embeddings` to `WhisperForAudioClassification`

younesbelkada opened the pull request:
# What does this PR do?

Fixes: https://github.com/huggingface/peft/issues/173

Fixes: https://github.com/huggingface/transformers/issues/22131

This PR adds `get_input_embeddings` method to `WhisperForAudioClassification` and `WhisperForConditionalGeneration` to avoid some issues such as [here](https://github.com/huggingface/peft/issues/173)

In my understanding `get_input_embeddings` method should return the first module that converts the input to the first hidden states and not necessarly an `nn.Embedding` layer

cc @ArthurZucker @sgugger

younesbelkada committed:
add `get_input_embeddings` to `WhisperForAudioClassification`

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_22133). All of your documentation changes will be reflected on that endpoint.

sgugger reviewed (approved):
LGTM, but let's also wait for @ArthurZucker to review before merging.

amyeroberts reviewed (approved):
LGTM. Have you tested the code with peft and confirmed this resolves the issue?

younesbelkada commented:
Yes I can confirm the script provided by the user:
```python
from transformers import WhisperFeatureExtractor
from transformers import WhisperTokenizer
from transformers import WhisperProcessor
from transformers import WhisperForAudioClassification

# Select CUDA device index
import os

os.environ["CUDA_VISIBLE_DEVICES"] = "0"
model_name_or_path = "openai/whisper-small"

feature_extractor = WhisperFeatureExtractor.from_pretrained(model_name_or_path)
tokenizer = WhisperTokenizer.from_pretrained(model_name_or_path)
processor = WhisperProcessor.from_pretrained(model_name_or_path)

model = WhisperForAudioClassification.from_pretrained(model_name_or_path, load_in_8bit=True, device_map="auto")

from peft import prepare_model_for_int8_training
model = prepare_model_for_int8_training(model)
```
work perfect now with this fix!

ArthurZucker commented:
I am a bit surprised by the fix, as it is not an embedding layer, this kind of break the usages we have for `get_input_embeddings`, which are for example `_resize_token_embeddings` and `tie_weights` which are both incompatible with the Whisper encoder as it does not have an embedding layer. 
So not really sure this is the way to go.
Also, the `test_model_common_attributes` has to be updated
```python 
    # WhisperEncoder has no inputs_embeds and thus the `get_input_embeddings` fn is not implemented
    def test_model_common_attributes(self):
        pass
```
as well as :
```python 
    # input embeds is meaningless for an encoder-only acoustic model
    def test_inputs_embeds(self):
        pass
```
in whisper tests

sgugger commented:
Even if it's not text, the layer converts an input to a hidden size, so it's kind of like an embedding. Not sure if there is a better name for this function but it doesn't shock me to use this for Whisper, esp since peft needs a common API across models.

As for the potential problems in `_resize_token_embeddings` and `tie_weights` I would wait for a user to actually raise an issue on this before giving them more thought. We can always override and add specific error messages.

younesbelkada committed:
add common tests

younesbelkada committed:
fix another common test

younesbelkada commented:
Added the common tests and there was an issue when running whisper without `decoder_input_ids` and `decoder_embeds` instead that I fixed. Can you please double check ? Thanks

ArthurZucker commented on src/transformers/models/whisper/modeling_whisper.py:
```diff
@@ -1029,7 +1029,10 @@ def forward(
 
         # embed positions
-        positions = self.embed_positions(input_ids, past_key_values_length=past_key_values_length)
+        if input_ids is not None:
+            positions = self.embed_positions(input_ids, past_key_values_length=past_key_values_length)
+        else:
+            positions = self.embed_positions(inputs_embeds, past_key_values_length=past_key_values_length)
```
Nice catch here! Seems like the input embeds were not possible to use

ArthurZucker commented on tests/models/whisper/test_modeling_whisper.py:
```diff
@@ -357,9 +357,26 @@ def _get_input_ids_and_config(self):
+            inputs = copy.deepcopy(self._prepare_for_class(inputs_dict, model_class))
+
+            encoder_input_ids = inputs["input_features"]
+            decoder_input_ids = inputs.get("decoder_input_ids", encoder_input_ids)
+            inputs.pop("decoder_input_ids", None)
```
```suggestion
decoder_input_ids = inputs.pop("decoder_input_ids", None)
```
Does this also work, feels a bit strange to default the `decoder_input_ids` to encoder's no? (small nit)

ArthurZucker reviewed (approved):
Thanks for the catch! LGTM

younesbelkada committed:
Update tests/models/whisper/test_modeling_whisper.py

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

younesbelkada committed:
fix style
